drivers/gpio/gpio_keys_zephyr.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio_keys.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(zephyr_gpio_keys, LOG_LEVEL_DBG);
+
+#define DT_DRV_COMPAT zephyr_gpio_keys
+
+struct gpio_keys_config {
+	/** GPIO specification from devicetree */
+	struct gpio_dt_spec spec;
+	/** Debounce interval in milliseconds from devicetree */
+	uint32_t debounce_interval_ms;
+	/** Zephyr code from devicetree */
+	uint32_t zephyr_code;
+};
+
+struct gpio_keys_pin_data {
+	const struct device *dev;
+	struct gpio_keys_callback cb_data;
+	struct k_work_delayable work;
+	int8_t pin_state;
+};
+
+struct gpio_keys_data {
+	gpio_keys_callback_handler_t callback;
+	int num_keys;
+	struct gpio_keys_pin_data *keys;
+};
+
+/**
+ * Handle debounced gpio pin state.
+ */
+static void gpio_keys_change_deferred(struct k_work *work)
+{
+	struct gpio_keys_pin_data *pin_data = CONTAINER_OF(work, struct gpio_keys_pin_data, work);
+	const struct device *dev = pin_data->dev;
+	struct gpio_keys_data *data = dev->data;
+	int key_index = pin_data - &data->keys[0];
+	const struct gpio_keys_config *cfg = &((struct gpio_keys_config *)dev->config)[key_index];
+
+	const int new_pressed = gpio_pin_get_raw(cfg->spec.port, cfg->spec.pin);
+
+	LOG_DBG("gpio_change_deferred %s pin_state=%d, new_pressed=%d, key_index=%d", dev->name,
+		pin_data->pin_state, new_pressed, key_index);
+
+	/* If gpio changed, invoke callback */
+	if (new_pressed != pin_data->pin_state) {
+		pin_data->pin_state = new_pressed;
+		LOG_DBG("Calling callback %s %d, code=%d", dev->name, new_pressed,
+			cfg->zephyr_code);
+		data->callback(dev, &pin_data->cb_data, BIT(cfg->spec.pin));
+	}
+}
+
+static void gpio_keys_change_call_deferred(struct gpio_keys_pin_data *data, uint32_t msec)
+{
+	int rv = k_work_reschedule(&data->work, K_MSEC(msec));
+
+	__ASSERT(rv >= 0, "Set wake mask work queue error");
+}
+
+static void gpio_keys_interrupt(const struct device *dev, struct gpio_callback *cbdata,
+				uint32_t pins)
+{
+	ARG_UNUSED(dev); /* This is a pointer to GPIO device, use dev pointer in
+			  * cbdata for pointer to gpio_debounce device node
+			  */
+	struct gpio_keys_pin_data *pin_data =
+		CONTAINER_OF(cbdata, struct gpio_keys_pin_data, cb_data);
+	const struct gpio_keys_config *cfg = pin_data->dev->config;
+	struct gpio_keys_data *data = pin_data->dev->data;
+
+	for (int i = 0; i < data->num_keys; i++) {
+		if (pins & BIT(cfg[i].spec.pin)) {
+			gpio_keys_change_call_deferred(pin_data, cfg->debounce_interval_ms);
+		}
+	}
+}
+
+static int enable_interrupt(const struct gpio_dt_spec *gpio_spec, struct gpio_keys_callback *cb,
```

Comment:
Rename to `gpio_keys_enable_interrupt`? As it is it somewhat looks like it wants to be a global function.

Reply:
done

Reply:
Oh I see it would collide with the API function. I think `_` prefixes should be avoided though, how about `gpio_keys_enable_interrupt_internal`?

Reply:
Renamed to `gpio_keys_interrupt_configure` to help align with GPIO pin API name.